pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java

```diff
@@ -216,18 +234,26 @@ public void channelRead(final ChannelHandlerContext ctx, Object msg) throws Exce
             break;
 
         case ProxyConnectionToBroker:
-            // Pass the buffer to the outbound connection and schedule next read
-            // only if we can write on the connection
-            ProxyService.OPS_COUNTER.inc();
-            if (msg instanceof ByteBuf) {
-                int bytes = ((ByteBuf) msg).readableBytes();
-                directProxyHandler.getInboundChannelRequestsRate().recordEvent(bytes);
-                ProxyService.BYTES_COUNTER.inc(bytes);
+            if (directProxyHandler != null) {
+                ProxyService.OPS_COUNTER.inc();
+                if (msg instanceof ByteBuf) {
+                    int bytes = ((ByteBuf) msg).readableBytes();
+                    directProxyHandler.getInboundChannelRequestsRate().recordEvent(bytes);
+                    ProxyService.BYTES_COUNTER.inc(bytes);
+                }
+                directProxyHandler.outboundChannel.writeAndFlush(msg)
+                        .addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE);
+            } else {
+                LOG.warn("Received message of type {} while connection to broker is missing in state {}. "
+                                + "Dropping the input message (readable bytes={}).", msg.getClass(), state,
+                        msg instanceof ByteBuf ? ((ByteBuf) msg).readableBytes() : -1);
             }
-            directProxyHandler.outboundChannel.writeAndFlush(msg)
-                    .addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE);
             break;
-
+        case ProxyConnectingToBroker:
+            LOG.warn("Received message of type {} while connecting to broker. "
+                            + "Dropping the input message (readable bytes={}).", msg.getClass(),
+                    msg instanceof ByteBuf ? ((ByteBuf) msg).readableBytes() : -1);
```

Comment:
Depending on the type of protocol message, dropping it could be problematic. For example, if the protocol message is a delivered message, it could lead to out of order persistence, as we saw in the broker.

Reply:
I'm assuming that the client is misbehaving if it's sending messages before it receives a response. Pulsar binary protocol spec about connecting: https://pulsar.apache.org/docs/en/develop-binary-protocol/#connection-establishment

Reply:
Yes, it would be misbehaving in that case. In the broker, we close the connection if a `send` message is sent before the producer is created successfully because otherwise, messages could be persisted out of order.

Reply:
In this case it's better to see if this warning gets ever logged. There's a possibility that some older Go client versions misbehave in some cases and disconnecting them would break the connectivity completely. Logging will reveal if that happens.

Reply:
Disconnecting might be the correct solution for handling misbehaving clients eventually.

Reply:
Okay. If we find that the go client is misbehaving, it'd be ideal to get a minor version released of the go client fixing that issue (I don't believe we currently do that for the go client, but implementing the protocol correctly is essential).

Also, I realize now that this kind of misbehaving is a bit different than the one we found in the broker. That one had to do with sending messages before the producer's creation finished. In this case, the dropped protocol message would be a create producer or consumer message, which is not as bad.